src/hdmint/tracker.cpp

```diff
@@ -475,85 +476,50 @@ bool CHDMintTracker::UpdateStatusInternal(const std::set<uint256>& setMempool, C
-    // Only allow updates for one or the other
-    if(fReset && fUpdateStatus)
-        return false;
-
-    std::list<CSigmaEntry> listMintsDB;
-    CWalletDB walletdb(strWalletFile);
-    walletdb.ListSigmaPubCoin(listMintsDB);
-    for (auto& mint : listMintsDB){
-        if(fReset){
-            mint.nHeight = -1;
-            mint.IsUsed = false;
-        }
-        else if(fUpdateStatus){
-            mint.IsUsed = fStatus;
-        }
-
-        if((!fSelection) ||
-            (fSelection && (serialHashes.find(GetSerialHash(mint.serialNumber)) != serialHashes.end()))){
-            Add(mint);
-        }
-    }
-    std::list<CHDMint> listDeterministicDB = walletdb.ListHDMints();
-
-    CHDMintWallet* zerocoinWallet = new CHDMintWallet(strWalletFile);
-    for (auto& dMint : listDeterministicDB) {
-        if(fReset){
-            dMint.SetHeight(-1);
-            dMint.SetUsed(false);
-        }
-        else if(fUpdateStatus){
-            dMint.SetUsed(fStatus);
-        }
-
-        if((!fSelection) ||
-            (fSelection && (serialHashes.find(dMint.GetSerialHash()) != serialHashes.end()))){
-            Add(dMint, true, false, zerocoinWallet);
-        }
-    }
-    delete zerocoinWallet;
-
-    return true;
-}
-
 void CHDMintTracker::UpdateMintStateFromBlock(const std::vector<sigma::PublicCoin>& mints){
+    CWalletDB walletdb(strWalletFile);
     std::vector<CMintMeta> updatedMeta;
+    std::list<std::pair<uint256, MintPoolEntry>> mintPoolEntries;
+    uint160 hashSeedMasterEntry;
+    CKeyID seedId;
+    int32_t nCount;
     std::set<uint256> setMempool;
     {
         LOCK(mempool.cs);
         mempool.getTransactions(setMempool);
     }
     for (auto& mint : mints) {
         uint256 hashPubcoin = sigma::GetPubCoinValueHash(mint.getValue());
-        if(!HasPubcoinHash(hashPubcoin))
-            continue;
-        CMintMeta meta = GetMetaFromPubcoin(hashPubcoin);
-        if(UpdateStatusInternal(setMempool, meta)){
-            updatedMeta.emplace_back(meta);
+        // Check hashPubcoin in db
+        if(walletdb.ReadMintPoolPair(hashPubcoin, hashSeedMasterEntry, seedId, nCount)){
+            // If found in db but not in memory - this is likely a resync
+            if(!HasPubcoinHash(hashPubcoin)){
+                MintPoolEntry mintPoolEntry(hashSeedMasterEntry, seedId, nCount);
+                mintPoolEntries.push_back(std::make_pair(hashPubcoin, mintPoolEntry));
+                continue;
+            }
+            CMintMeta meta = GetMetaFromPubcoin(hashPubcoin);
+            if(UpdateMetaStatus(setMempool, meta)){
+                updatedMeta.emplace_back(meta);
+            }
         }
     }
 
+    if(mintPoolEntries.size()>0)
+        zwalletMain->SyncWithChain(false, mintPoolEntries);
+
     //overwrite any updates
     for (CMintMeta& meta : updatedMeta)
         UpdateState(meta);
 }
 
 void CHDMintTracker::UpdateSpendStateFromBlock(const sigma::spend_info_container& spentSerials){
```

Comment:
I see we have duplication of code in this 2 functions. Would be better somehow prevent that.

Reply:
Created new function `UpdateFromBlock`